File: inJava/app/src/main/java/com/barrygithub/rickandmortyapp/ui/viewModel/ViewModelMain.java
package com.barrygithub.rickandmortyapp.ui.viewModel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.EntityDb;
import com.barrygithub.rickandmortyapp.data.RemoteDatasource.entities.Episode;
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractor;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.core.SingleObserver;
import io.reactivex.rxjava3.disposables.Disposable;

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 3/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 **/
@HiltViewModel
public class ViewModelMain extends ViewModel {
    private UseCaseInteractor interactor;
    public Disposable disposable;

    private MutableLiveData<EntityDb> _responseApi = new MutableLiveData<>();
    public LiveData<EntityDb> getResponseApi() {return _responseApi;}
    private MutableLiveData<Boolean> _loading = new MutableLiveData<>(true);
    public LiveData<Boolean> isLoading(){return _loading;}
    private MutableLiveData<Boolean> _loadingEpisode = new MutableLiveData<>(true);
    public LiveData<Boolean> isLoadingEpisode(){return _loadingEpisode;}

    private MutableLiveData<Episode> _episode = new MutableLiveData<>();
    public LiveData<Episode> getEpisode(){return _episode;}
    private MutableLiveData<String> _error= new MutableLiveData<>();
    public LiveData<String> getError(){return _error;}

  public @Inject ViewModelMain(UseCaseInteractor interactor){
      this.interactor= interactor;
      observerData();
  }
  public void getData(int page){
        interactor.getData(page);
  }
    private void observerData(){
        Observable<EntityDb> response= interactor.observerData();
        response.observeOn(AndroidSchedulers.mainThread())
                .subscribe(getResponse());
    }

   private Observer<EntityDb> getResponse(){
        return new Observer<EntityDb>() {
            @Override
            public void onSubscribe(@NonNull Disposable d) {
                disposable = d;
            }

            @Override
            public void onNext(@NonNull EntityDb entitydb) {
                _responseApi.postValue(entitydb);
                _loading.postValue(false);

            }

            @Override
            public void onError(@NonNull Throwable e) {
                _loading.postValue(false);
                _responseApi.postValue(null);
                _error.postValue(e.getMessage());

                Log.e("TAG", "Error de conexión");

            }

            @Override
            public void onComplete() {
                _loading.postValue(false);
            }
        };
    }
    public void callEpisode(int isEpisode){
        interactor.getEpisodeFromApi(isEpisode)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<Episode>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                        disposable=d;
                    }

                    @Override
                    public void onSuccess(@NonNull Episode episode) {
                        _episode.postValue(episode);
                        _loadingEpisode.postValue(false);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        Log.e("TAG", e.getMessage());

                    }
                });
    }

}

File: inKotlin/app/src/main/java/com/barrygithub/rickandmortyapp/ui/viewModel/ViewModelMain.kt
package com.barrygithub.rickandmortyapp.ui.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barrygithub.rickandmortyapp.data.localDatasource.EntityDb
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractor
import dagger.hilt.android.lifecycle.HiltViewModel
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 14/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 *
 **/
@HiltViewModel
class ViewModelMain @Inject constructor(private val interactor: UseCaseInteractor):ViewModel() {

    lateinit var disposable:Disposable
    private var _responseApi:MutableLiveData<EntityDb> = MutableLiveData()
    val responseApi:LiveData<EntityDb> get() = _responseApi

    private var _isLoading:MutableLiveData<Boolean> = MutableLiveData(true)
    val isLoading:LiveData<Boolean> get() = _isLoading

    private var _loadingEpisode:MutableLiveData<Boolean> = MutableLiveData(true)
    val loadingEpisode:LiveData<Boolean> get() = _loadingEpisode



    private var _error:MutableLiveData<String> = MutableLiveData()
    val error:LiveData<String> get() = _error

    fun getData(page: Int){
        viewModelScope.launch {
            interactor.getDataGraphql(page)
        }
    }
    init {
        setUpObserverData()
    }

    private fun setUpObserverData(){
        var response = interactor.getObservableData()
        response.observeOn(AndroidSchedulers.mainThread())
            .subscribe(getResponse())
    }

    private fun getResponse(): Observer<EntityDb> {
        return object : Observer<EntityDb>{
            override fun onSubscribe(d: Disposable){disposable=d}
            override fun onNext(response: EntityDb) {
                _responseApi.postValue(response)
                _isLoading.postValue(false)

            }

            override fun onError(e: Throwable) {
               _isLoading.postValue(false)
                _responseApi.postValue(EntityDb())
                _error.postValue(e.message)
            }

            override fun onComplete() {
                _isLoading.postValue(false)
            }
        }
    }

}